cmd/secrets/providers/providers.go

```diff
@@ -0,0 +1,68 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// +build secrets
+
+package providers
+
+import (
+	"fmt"
+	"time"
+
+	s "github.com/DataDog/datadog-agent/pkg/secrets"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver"
+)
+
+type SecretsProvider interface {
+	// Get Fetches a secret
+	// Note that secretID has a different format on each secret provider. In the
+	// file provider, it's a path, whereas in the kubernetes secret one it has
+	// the following format: "namespace/name/key".
+	Get(secretID string) s.Secret
+}
+
+type Prefix string
+
+const (
+	file      Prefix = "file"
+	k8sSecret Prefix = "k8s_secret"
+)
+
+func (prefix Prefix) IsValid() bool {
+	return prefix == file || prefix == k8sSecret
+}
+
+type Selector struct {
+	File      *FileProvider
+	K8sSecret *K8sSecretProvider
+}
+
+func (selector *Selector) GetProvider(prefix Prefix) (SecretsProvider, error) {
```

Comment:
Same here, I feel like a simple `GetSecretFromProviders` function would be better, we don't need the `Selector` struct. Which would also simplify the `readSecrets` interface.

Unless `GetKubeClient` is very resource intensive we can afford to get a new client for each secret (I think). There will not be hundreds of secrets per configuration.

Reply:
See comment above.

Reply:
I guess you can have a cached client map that you check for the provider. If the provider isn't there, instantiate it. Probably would be in the realm of over-engineering though.

```golang
var  providers map[Prefix]SecretsProvider
...

init() {
    providers = make(map[Prefix]SecretsProvider)
}
...

func() GetSecret(prefix Prefix, secretID string) (s.Secret, error) {
    ...

    if provider, ok = providers[prefix]; !ok {
        // create the prefixed provider
    }
    ...    

    return provider.Get(secretID)
} 
```

Either way, I don't feel strongly enough about this on my end.

Reply:
I'm not sure about this. I think it would be difficult to use it in the `secret_helper_test.go`.